### PR TITLE
[export] Improve consistency for nn_module_stack metadata, add checks to _trace.py

### PR DESCRIPTION
**Pull request by pianpwk**
We would like to improve consistency for nn_module_stack metadata in torch.export.

This PR ensures that all tests in test/export/test_export.py has the following constraints:
- Remove nn_module_stack for all placeholder & output nodes, for all modules and submodules
- Ensure nn_module_stack is present for all other node types for the top-level module (there is still an issue with torch.cond submodules having empty fields)
- Add these checks to _export() in _trace.py (we would add this in the Verifier, but downstream apps construct ExportedPrograms separate from _export(), and metadata may not be maintained there)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

File: torch/_export/verifier.py [hunk 1]
```diff
@@ -31,6 +31,35 @@ def _check_has_fake_tensor(node: torch.fx.Node) -> None:
...
+        - None for 'placeholder', output'
+
+    TODO(pianpwk): extend this check to get_attr & call_module nodes.
+    TODO(pianpwk): make this a consistent node-level check once nn_module_stack is populated for cond submodules.
+    '''
+    # Check top-level graph for all nodes, submodule graphs for placeholder & output nodes
+    for i, mod in enumerate(graph_module.modules()):
+        for node in mod.graph.nodes:
+            if node.op in ['call_method', 'call_function', 'get_attr', 'call_module']:
+                if i == 0:
```
Comment (angelayi) [inline torch/_export/verifier.py:51]: hmm, I'm not sure if the ordering is consistent in the `.modules()` calls, where i=0 is always the toplevel module. Maybe it's better if you first run a check on `graph_module` call_function nodes, and then do loop over all the modules to check for the placeholder/output nodes.

File: torch/_export/verifier.py [hunk 2]
```diff
@@ -31,6 +31,35 @@ def _check_has_fake_tensor(node: torch.fx.Node) -> None:
     return _check_val(node)
 
 
+def check_nn_module_stack(graph_module: torch.fx.GraphModule) -> None:
```
Comment (zhxchen17) [inline torch/_export/verifier.py:34]: Could we also call this function directly inside verifier?

File: test/export/test_export_nn_module_stack.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    import test_export
+    import testing
+from torch._export.verifier import check_nn_module_stack
+from torch.export import export
+
+test_classes = {}
+
+"""
+Ideally this test file would not exist, and torch._export.verifier would call check_nn_module_stack() in its Verifier class.
+For now nn_module_stack consistency is not fully covered (e.g. (de)serialization, export passes),
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply test/export/test_export_nn_module_stack.py:15]: Yeah, for serialization I think it's calls like this, I ran into issues with for test_serialize.py (deserializing sym_int nodes): https://github.com/pytorch/pytorch/blob/f94933ed42be213d37032186e9465b641ef00585/torch/_export/serde/serialize.py#L1358-L1363

or this (deserializing get_attr nodes for submodules): https://github.com/pytorch/pytorch/blob/f94933ed42be213d37032186e9465b641ef00585/torch/_export/serde/serialize.py#L1576-L1580

For passes one example I ran into was this: https://github.com/pytorch/pytorch/blob/67d3e4f2a298f89cbe3b5ecc304bc45011a6ce09/torch/_export/passes/replace_set_grad_with_hop_pass.py#L54-L56

I tried to start fixing them but seems like there's many other sources, so I was hoping to lower the scope of this PR

File: torch/_export/verifier.py (hunk 2 above)
Comment (pianpwk) [reply torch/_export/verifier.py:34]: That was the original idea/implementation, but re. the comments in test/export/test_export_nn_module_stack.py, this check still fails with tests in serialization/export passes :(

I was hoping to add that file to add verification onto tests in test_export.py for now, but agree that this isn't very elegant.

File: torch/_export/verifier.py [hunk 4]
```diff
@@ -31,6 +31,35 @@ def _check_has_fake_tensor(node: torch.fx.Node) -> None:
...
+    '''
+    Perform nn_module_stack checks on the graph.
+    Current constraints:
+        For the top level graph:
+        - populated for 'call_method', 'call_function', 'get_attr', 'call_module'
+        - None for 'placeholder', 'output'
+        For submodule graphs:
+        - None for 'placeholder', output'
+
+    TODO(pianpwk): extend this check to get_attr & call_module nodes.
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/_export/verifier.py:44]: Ah gotcha, just changed!

File: torch/_export/verifier.py (hunk 1 above)
Comment (pianpwk) [reply torch/_export/verifier.py:51]: Good point, thank you! Just changed :)

File: torch/export/_trace.py [hunk 5]
```diff
@@ -948,6 +954,10 @@ def forward(self, *args, **kwargs):
     export_graph_signature = ep_non_strict.sig
     constants = ep_non_strict.constants
 
+    # Don't copy over nn_module_stack metadata for params/buffers nodes
+    for metadata in params_buffers_to_node_meta.values():
+        metadata.pop("nn_module_stack", None)
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/export/_trace.py:959]: Ahh I think that code only applies to non_strict mode, this is metadata that's been lifted in strict mode. The non-strict case should be handled by the changes in L448-453.

File: torch/fx/experimental/proxy_tensor.py [hunk 6]
```diff
@@ -979,6 +979,16 @@ def call_module(self, m, forward, args, kwargs):
     def is_leaf_module(self, m, module_qualified_name):
         return False
 
+    def create_node(self, *args, **kwargs):
+        '''
+        Add nn_module_stack metadata here instead of TracerBase,
+        since calls to make_fx() might not want to record module stack metadata
```
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/fx/experimental/proxy_tensor.py:985]: Yeah the latter case is right! There's calls like make_fx(record_module_stack=False) that want to go through PythonKeyTracer instead, so I chose to patch it here. Maybe there's a better way though?

File: torch/_export/verifier.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/_export/verifier.py:44]: maybe edit this comment too?

File: torch/fx/experimental/proxy_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply torch/fx/experimental/proxy_tensor.py:985]: Ah no that's fine, I think the comment is just a little confusing to me. Maybe you can just say something like "Patch the nn_module_stack if TracerBase did not add it"

File: test/export/test_export_nn_module_stack.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (angelayi) [reply test/export/test_export_nn_module_stack.py:15]: Got it, thanks for the pointers!

**Review by pianpwk — COMMENTED** [review]
clicked this by mistake

File: torch/fx/experimental/proxy_tensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pianpwk) [reply torch/fx/experimental/proxy_tensor.py:985]: Ah thank you! Will add this soon

File: test/export/test_export_nn_module_stack.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply test/export/test_export_nn_module_stack.py:15]: Hmm, it's a bit hard to understand where we stand after this PR until you fix these issues that came up (or record them as expected failures).

**Review by avikchaudhuri — APPROVED** [review]
Accepting to unblock but this seems like unfinished business. Can you update the PR with the follow-up plan?

**Comment by avikchaudhuri** [comment]
Chatted offline with @pianpwk, plan of record is to introduce expected failures under a new decorator (assuming it's a small number of cases and we've done due diligence to ensure that the invariant is not unreasonable). Then we track those as part of the overall hardening plan and parallelize fixing those cases.

**Comment by pianpwk** [comment]
Updated PR to fix failures in export passes/deserialization/tp_transform and add checks to verifier. 1 failure remaining in test_serialize.py regarding sym_size_int nodes.

File: torch/_export/serde/serialize.py [hunk 7]
```diff
@@ -1380,6 +1380,15 @@ def deserialize_node(self, serialized_node: Node, target: Callable) -> None:
             self.deserialize_outputs(serialized_node, fx_node)
             fx_node.meta.update(self.deserialize_metadata(serialized_node.metadata))
 
+            # assign nn_module_stack metadata to submodule/subgraph nodes for higher order ops
+            if serialized_node.target == "torch.ops.higher_order.cond":
```
Comment (zhxchen17) [inline torch/_export/serde/serialize.py:1384]: I don't think it's very generic to enumerate all the higher order ops here. If we're adding other HOOs in the future, we could fail silently here.

I think it's okay to loop over all the graph arguments and not special casing on these ops, maybe just use "torch.ops.higher_order" as common prefix to test.

File: torch/_export/verifier.py [hunk 8]
```diff
@@ -248,6 +276,7 @@ def _is_type(name, ty):
                 # elif node.op == "output":
                 #     _check_flattened_outputs()
 
+        check_nn_module_stack(gm)
```
Comment (zhxchen17) [inline torch/_export/verifier.py:279]: nice

**Review by zhxchen17 — CHANGES_REQUESTED** [review]

File: torch/_export/serde/serialize.py (hunk 7 above)
Comment (pianpwk) [reply torch/_export/serde/serialize.py:1384]: Thanks! That idea makes sense.

Comment (pianpwk) [reply torch/_export/serde/serialize.py]: @zhxchen17 could you help re-review? Just made the change.

**Review by avikchaudhuri — COMMENTED** [review]
looks great

**Pull request closed without merging**